File: server/riverbed/store.py
"""In-memory columnar ring buffer, plus sketch-backed long-horizon rollups.

Why not a database: every query here is a full scan over a bounded, recent
time range with no point lookups and no joins. An embedded database would add
serialization, a query planner fighting ours, and disk I/O to a workload that
fits comfortably in RAM. A ring buffer of parallel column arrays is both faster
and much easier to reason about for continuous re-evaluation.

Why columnar: a typical query touches two or three of fifteen columns. Column
arrays mean those scans walk contiguous memory instead of skipping through
whole event objects, and Python's list-of-primitives is dramatically cheaper to
iterate than a list of dicts or dataclasses.
"""

from __future__ import annotations

import threading
import time
from dataclasses import dataclass, field
from typing import Any, Iterable

from .schema import COLUMNS, ColType
from .sketches import CountMinSketch, HyperLogLog, TDigest

#: Raw retention. Sized so the whole buffer stays a few hundred MB at the
#: firehose's real rate (~40 events/sec -> ~72k rows per 30 min).
DEFAULT_WINDOW_SECONDS = 30 * 60

#: Hard cap so a traffic spike cannot exhaust memory. Whichever bound is hit
#: first wins.
DEFAULT_MAX_ROWS = 250_000


@dataclass
class RollupBucket:
    """One minute of pre-aggregated history, for horizons past the raw buffer."""

    minute: int
    events: int = 0
    bot_events: int = 0
    anon_events: int = 0
    bytes_added: int = 0
    bytes_removed: int = 0
    users: HyperLogLog = field(default_factory=lambda: HyperLogLog(12))
    titles: CountMinSketch = field(default_factory=lambda: CountMinSketch(width=1024, depth=4, k=40))
    deltas: TDigest = field(default_factory=TDigest)
    by_lang: dict[str, int] = field(default_factory=dict)

    def to_dict(self) -> dict:
        return {
            "minute": self.minute,
            "events": self.events,
            "bot_events": self.bot_events,
            "anon_events": self.anon_events,
            "bytes_added": self.bytes_added,
            "bytes_removed": self.bytes_removed,
            "distinct_users": self.users.count(),
            "p50_delta": round(self.deltas.quantile(0.5), 1),
            "p99_delta": round(self.deltas.quantile(0.99), 1),
        }


class EventStore:
    """Thread-safe columnar ring buffer.

    Locking is coarse -- one lock around append and snapshot -- because the
    contention pattern is one writer (the ingest task) and short reads from the
    query loop. Finer-grained locking would buy nothing and risks tearing a
    scan across an eviction.
    """

    def __init__(
        self,
        window_seconds: int = DEFAULT_WINDOW_SECONDS,
        max_rows: int = DEFAULT_MAX_ROWS,
        rollup_minutes: int = 24 * 60,
    ):
        self.window_seconds = window_seconds
        self.max_rows = max_rows
        self.rollup_minutes = rollup_minutes

        self._lock = threading.RLock()
        self.columns: dict[str, list[Any]] = {c.name: [] for c in COLUMNS}
        self._col_types = {c.name: c.type for c in COLUMNS}

        #: Monotonically increasing id of the oldest retained row. Lets the API
        #: tell a client "you missed rows" instead of silently skipping.
        self.base_offset = 0
        self.total_ingested = 0

        self.rollups: dict[int, RollupBucket] = {}
        self.started_at = time.time()
        self.last_event_at: float | None = None

    # -- writing ------------------------------------------------------------

    def append(self, row: dict[str, Any]) -> None:
        with self._lock:
            for name, col in self.columns.items():
                col.append(row.get(name, _zero(self._col_types[name])))
            self.total_ingested += 1
            self.last_event_at = time.time()
            self._update_rollup(row)
            self._evict()

    def append_many(self, rows: Iterable[dict[str, Any]]) -> int:
        count = 0
        with self._lock:
            for row in rows:
                for name, col in self.columns.items():
                    col.append(row.get(name, _zero(self._col_types[name])))
                self._update_rollup(row)
                count += 1
            self.total_ingested += count
            if count:
                self.last_event_at = time.time()
            self._evict()
        return count

    def _update_rollup(self, row: dict[str, Any]) -> None:
        minute = int(row.get("ts", time.time())) // 60
        bucket = self.rollups.get(minute)
        if bucket is None:
            bucket = self.rollups[minute] = RollupBucket(minute=minute)

        bucket.events += 1
        if row.get("is_bot"):
            bucket.bot_events += 1
        if row.get("is_anon"):
            bucket.anon_events += 1

        delta = row.get("delta", 0) or 0
        if delta > 0:
            bucket.bytes_added += delta
        else:
            bucket.bytes_removed += -delta
        bucket.deltas.add(float(delta))

        if user := row.get("user"):
            bucket.users.add(user)
        if title := row.get("title"):
            bucket.titles.add(title)
        if lang := row.get("lang"):
            bucket.by_lang[lang] = bucket.by_lang.get(lang, 0) + 1

        if len(self.rollups) > self.rollup_minutes:
            for stale in sorted(self.rollups)[: len(self.rollups) - self.rollup_minutes]:
                del self.rollups[stale]

    def _evict(self) -> None:
        """Drop rows that fell out of the retention window or the row cap.

        Eviction is a single slice per column rather than repeated pops: slicing
        is one C-level memmove, while popping from the front of a Python list is
        O(n) *per row* and would dominate the ingest path under load.
        """
        n = len(self.columns["ts"])
        if n == 0:
            return

        cutoff = time.time() - self.window_seconds
        ts = self.columns["ts"]

        # Rows arrive in near-timestamp order, so a linear scan from the front
        # is effectively O(evicted) rather than O(n).
        drop = 0
        while drop < n and ts[drop] < cutoff:
            drop += 1

        if n - drop > self.max_rows:
            drop = n - self.max_rows

        if drop:
            for name in self.columns:
                del self.columns[name][:drop]
            self.base_offset += drop

    # -- reading ------------------------------------------------------------

    def snapshot(self, window_seconds: int | None = None) -> "Scan":
        """Return an immutable view of the rows inside `window_seconds`.

        The column lists are sliced (copied) under the lock so the executor can
        scan without holding it. At these sizes the copy is far cheaper than
        blocking ingest for the duration of a query.
        """
        with self._lock:
            n = len(self.columns["ts"])
            if n == 0:
                return Scan({name: [] for name in self.columns}, 0)

            start = 0
            if window_seconds is not None:
                cutoff = time.time() - window_seconds
                ts = self.columns["ts"]
                # Binary search: timestamps are effectively sorted.
                lo, hi = 0, n
                while lo < hi:
                    mid = (lo + hi) // 2
                    if ts[mid] < cutoff:
                        lo = mid + 1
                    else:
                        hi = mid
                start = lo

            return Scan({name: col[start:] for name, col in self.columns.items()}, n - start)

    def recent(self, limit: int = 50, human_only: bool = False) -> list[dict[str, Any]]:
        """Most recent rows, newest first -- powers the live event feed.

        `human_only` restricts to people editing real Wikipedia articles. The
        filter has to happen here rather than in the client: bots are ~58% of
        the stream and most of the rest is Wikidata, so filtering a batch of a
        dozen already-sent rows would usually leave one or two.

        The scan walks backwards and stops as soon as it has enough, so the cost
        is proportional to what is returned, not to the buffer size.
        """
        with self._lock:
            n = len(self.columns["ts"])
            cols = self.columns
            rows: list[dict[str, Any]] = []

            for i in range(n - 1, -1, -1):
                if human_only and (
                    cols["is_bot"][i]
                    or cols["project"][i] != "wikipedia"
                    or cols["namespace"][i] != 0
                ):
                    continue
                rows.append({name: col[i] for name, col in cols.items()})
                if len(rows) >= limit:
                    break

        return rows

    def stats(self) -> dict[str, Any]:
        with self._lock:
            n = len(self.columns["ts"])
            now = time.time()
            recent_cut = now - 60
            ts = self.columns["ts"]
            per_min = sum(1 for i in range(n - 1, -1, -1) if ts[i] >= recent_cut) if n else 0
            # Divide by elapsed time, not a flat 60s: for the first minute after
            # start there is less than a minute of data and a flat divisor
            # under-reports the rate by up to 60x.
            elapsed = max(1.0, min(60.0, now - self.started_at))
            return {
                "buffered_rows": n,
                "total_ingested": self.total_ingested,
                "events_per_second": round(per_min / elapsed, 2),
                "window_seconds": self.window_seconds,
                "rollup_minutes": len(self.rollups),
                "uptime_seconds": round(now - self.started_at, 1),
                "last_event_age": round(now - self.last_event_at, 2) if self.last_event_at else None,
            }

    def rollup_series(self, minutes: int = 180) -> list[dict]:
        with self._lock:
            keys = sorted(self.rollups)[-minutes:]
            return [self.rollups[k].to_dict() for k in keys]


class Scan:
    """An immutable columnar slice handed to the executor."""

    __slots__ = ("columns", "length")

    def __init__(self, columns: dict[str, list[Any]], length: int):
        self.columns = columns
        self.length = length

    def column(self, name: str) -> list[Any]:
        return self.columns[name]

    def __len__(self) -> int:
        return self.length


def _zero(t: ColType) -> Any:
    if t is ColType.STRING:
        return ""
    if t is ColType.BOOL:
        return False
    if t is ColType.FLOAT:
        return 0.0
    return 0

File: src/engine/store.ts
import { COLUMNS, type CellValue, type EditRow, zeroFor } from './schema';
import { CountMinSketch, HyperLogLog, TDigest } from './sketches';

/**
 * In-memory columnar ring buffer, plus sketch-backed rollups.
 *
 * Why not IndexedDB or a WASM database: every query here is a full scan over a
 * bounded, recent time range with no point lookups and no joins. A database
 * would add serialization, its own query planner fighting ours, and async I/O
 * to a workload that fits comfortably in a tab's memory.
 *
 * Why columnar: a typical query touches two or three of fifteen columns. Column
 * arrays mean those scans walk contiguous memory instead of chasing pointers
 * through per-row objects, and a JS engine can keep a monomorphic array of
 * numbers unboxed, which it cannot do for object properties.
 */

/**
 * Raw retention. Shorter than the server version's 30 minutes: this runs in a
 * visitor's tab alongside everything else they have open, and 15 minutes at the
 * real firehose rate is already ~20k rows.
 */
export const DEFAULT_WINDOW_SECONDS = 15 * 60;

/**
 * Hard row cap, so a traffic spike or a long-lived tab cannot exhaust memory.
 * Whichever bound is hit first wins.
 */
export const DEFAULT_MAX_ROWS = 80_000;

export interface RollupBucket {
  minute: number;
  events: number;
  botEvents: number;
  anonEvents: number;
  bytesAdded: number;
  bytesRemoved: number;
  users: HyperLogLog;
  titles: CountMinSketch;
  deltas: TDigest;
}

export interface RollupSummary {
  minute: number;
  events: number;
  botEvents: number;
  anonEvents: number;
  bytesAdded: number;
  bytesRemoved: number;
  distinctUsers: number;
  p50Delta: number;
  p99Delta: number;
}

export interface StoreStats {
  bufferedRows: number;
  totalIngested: number;
  eventsPerSecond: number;
  windowSeconds: number;
  oldestAge: number | null;
  lastEventAge: number | null;
  uptimeSeconds: number;
}

/** An immutable columnar slice handed to the executor. */
export interface Scan {
  columns: Record<string, CellValue[]>;
  length: number;
}

const now = () => Date.now() / 1000;

export class EventStore {
  readonly windowSeconds: number;
  readonly maxRows: number;
  readonly columns: Record<string, CellValue[]> = {};

  private readonly rollupMinutes: number;
  private readonly rollups = new Map<number, RollupBucket>();
  private readonly startedAt = now();

  /** Oldest retained row id, so callers can tell rows were dropped. */
  baseOffset = 0;
  totalIngested = 0;
  lastEventAt: number | null = null;

  constructor(
    windowSeconds = DEFAULT_WINDOW_SECONDS,
    maxRows = DEFAULT_MAX_ROWS,
    rollupMinutes = 180,
  ) {
    this.windowSeconds = windowSeconds;
    this.maxRows = maxRows;
    this.rollupMinutes = rollupMinutes;
    for (const column of COLUMNS) this.columns[column.name] = [];
  }

  appendMany(rows: EditRow[]): number {
    if (rows.length === 0) return 0;

    for (const row of rows) {
      for (const column of COLUMNS) {
        const value = row[column.name];
        this.columns[column.name]!.push(value === undefined ? zeroFor(column.type) : value);
      }
      this.updateRollup(row);
    }

    this.totalIngested += rows.length;
    this.lastEventAt = now();
    this.evict();
    return rows.length;
  }

  append(row: EditRow): void {
    this.appendMany([row]);
  }

  private updateRollup(row: EditRow): void {
    const minute = Math.floor(Number(row.ts ?? now()) / 60);
    let bucket = this.rollups.get(minute);
    if (bucket === undefined) {
      bucket = {
        minute,
        events: 0,
        botEvents: 0,
        anonEvents: 0,
        bytesAdded: 0,
        bytesRemoved: 0,
        users: new HyperLogLog(11),
        titles: new CountMinSketch(1024, 4, 40),
        deltas: new TDigest(),
      };
      this.rollups.set(minute, bucket);
    }

    bucket.events += 1;
    if (row.is_bot) bucket.botEvents += 1;
    if (row.is_anon) bucket.anonEvents += 1;

    const delta = Number(row.delta ?? 0);
    if (delta > 0) bucket.bytesAdded += delta;
    else bucket.bytesRemoved += -delta;
    bucket.deltas.add(delta);

    if (row.user) bucket.users.add(String(row.user));
    if (row.title) bucket.titles.add(String(row.title));

    if (this.rollups.size > this.rollupMinutes) {
      const stale = [...this.rollups.keys()].sort((a, b) => a - b);
      for (let i = 0; i < stale.length - this.rollupMinutes; i += 1) {
        this.rollups.delete(stale[i]!);
      }
    }
  }

  /**
   * Drop rows past the retention window or the row cap.
   *
   * Eviction is a single splice per column rather than repeated shifts: shifting
   * from the front of an array is O(n) *per row* and would dominate the ingest
   * path, while one splice is a single bulk move.
   */
  private evict(): void {
    const ts = this.columns.ts!;
    const n = ts.length;
    if (n === 0) return;

    const cutoff = now() - this.windowSeconds;

    // Rows arrive in near-timestamp order, so scanning from the front costs
    // O(evicted) rather than O(n).
    let drop = 0;
    while (drop < n && Number(ts[drop]) < cutoff) drop += 1;

    if (n - drop > this.maxRows) drop = n - this.maxRows;

    if (drop > 0) {
      for (const column of COLUMNS) this.columns[column.name]!.splice(0, drop);
      this.baseOffset += drop;
    }
  }

  /**
   * An immutable view of the rows inside `windowSeconds`.
   *
   * Unlike the server version this does not copy: the worker is single-threaded
   * and nothing mutates the store while a query runs, so slicing would be pure
   * waste. A start offset is carried instead, and the executor indexes from it.
   */
  snapshot(windowSeconds: number | null): Scan {
    const ts = this.columns.ts!;
    const n = ts.length;
    if (n === 0) {
      const empty: Record<string, CellValue[]> = {};
      for (const column of COLUMNS) empty[column.name] = [];
      return { columns: empty, length: 0 };
    }

    let start = 0;
    if (windowSeconds !== null) {
      const cutoff = now() - windowSeconds;
      // Binary search: timestamps are effectively sorted.
      let lo = 0;
      let hi = n;
      while (lo < hi) {
        const mid = (lo + hi) >> 1;
        if (Number(ts[mid]) < cutoff) lo = mid + 1;
        else hi = mid;
      }
      start = lo;
    }

    if (start === 0) return { columns: this.columns, length: n };

    const sliced: Record<string, CellValue[]> = {};
    for (const column of COLUMNS) sliced[column.name] = this.columns[column.name]!.slice(start);
    return { columns: sliced, length: n - start };
  }

  /**
   * Most recent rows, newest first, for the live feed.
   *
   * `humanOnly` restricts to people editing real Wikipedia articles. The filter
   * belongs here rather than in the UI: bots are ~58% of the stream and most of
   * the rest is Wikidata, so filtering a dozen already-selected rows would
   * usually leave one or two. The scan walks backwards and stops as soon as it
   * has enough, so it costs what it returns, not the buffer size.
   */
  recent(limit = 20, humanOnly = false): EditRow[] {
    const n = this.columns.ts!.length;
    const rows: EditRow[] = [];

    for (let i = n - 1; i >= 0 && rows.length < limit; i -= 1) {
      if (humanOnly) {
        if (this.columns.is_bot![i]) continue;
        if (this.columns.project![i] !== 'wikipedia') continue;
        if (this.columns.namespace![i] !== 0) continue;
      }
      const row: EditRow = {};
      for (const column of COLUMNS) row[column.name] = this.columns[column.name]![i]!;
      rows.push(row);
    }

    return rows;
  }

  stats(): StoreStats {
    const ts = this.columns.ts!;
    const n = ts.length;
    const current = now();

    // Rate over the last minute of *event time*, divided by the span of event
    // time actually covered.
    //
    // Dividing by wall-clock elapsed instead would be wrong in both directions:
    // it under-reports by up to 60x during the first minute (less than a minute
    // of data, flat 60s divisor), and it wildly over-reports during the startup
    // backfill, where 90 seconds of history arrives in about two seconds.
    // Measuring event time against event time is immune to both.
    const recentCut = current - 60;
    let recentCount = 0;
    for (let i = n - 1; i >= 0; i -= 1) {
      if (Number(ts[i]) < recentCut) break;
      recentCount += 1;
    }
    const oldestRetained = n > 0 ? Number(ts[0]) : current;
    const covered = current - Math.max(oldestRetained, recentCut);
    const elapsed = Math.max(1, Math.min(60, covered));

    return {
      bufferedRows: n,
      totalIngested: this.totalIngested,
      eventsPerSecond: Number((recentCount / elapsed).toFixed(2)),
      windowSeconds: this.windowSeconds,
      oldestAge: n > 0 ? Number((current - Number(ts[0])).toFixed(1)) : null,
      lastEventAge: this.lastEventAt ? Number((current - this.lastEventAt).toFixed(2)) : null,
      uptimeSeconds: Number((current - this.startedAt).toFixed(1)),
    };
  }

  rollupSeries(minutes = 180): RollupSummary[] {
    const keys = [...this.rollups.keys()].sort((a, b) => a - b).slice(-minutes);
    return keys.map((key) => {
      const bucket = this.rollups.get(key)!;
      return {
        minute: bucket.minute,
        events: bucket.events,
        botEvents: bucket.botEvents,
        anonEvents: bucket.anonEvents,
        bytesAdded: bucket.bytesAdded,
        bytesRemoved: bucket.bytesRemoved,
        distinctUsers: bucket.users.count(),
        p50Delta: Number(bucket.deltas.quantile(0.5).toFixed(1)),
        p99Delta: Number(bucket.deltas.quantile(0.99).toFixed(1)),
      };
    });
  }
}
